platform/darwin/src/MGLStyleValue_Private.h

```
@@ -115,162 +117,14 @@ class MGLStyleValueTransformer {
-                    };
-                    if (compositeStyleFunction.defaultValue) {
-                        NSCAssert([compositeStyleFunction.defaultValue isKindOfClass:[MGLStyleConstantValue class]], @"Default value must be constant");
-                        MBGLType mbglValue;
-                        id mglValue = [(MGLStyleConstantValue<ObjCType> *)compositeStyleFunction.defaultValue rawValue];
-                        getMBGLValue(mglValue, mbglValue);
-                        compositeFunction.defaultValue = mbglValue;
-                    }
-                    return compositeFunction;
-                }
-                    break;
-                default:
-                    [NSException raise:NSInvalidArgumentException
-                                format:@"A composite function must use exponential, interval, or categorical stops."];
-                    return {};
-            }
-            return {};
-        } else if ([value isMemberOfClass:[MGLStyleFunction class]]) {
-            MGLStyleFunction<ObjCType> *styleFunction = (MGLStyleFunction<ObjCType> *)value;
-            return toMBGLExponentialCameraFunction(styleFunction);
-        } else if (value) {
-            [NSException raise:@"MGLAbstractClassException" format:
-             @"The style value %@ cannot be applied to the style. "
-             @"Make sure the style value was created as a member of a concrete subclass of MGLStyleValue.",
-             NSStringFromClass([value class])];
-            return {};
-        } else {
-            return {};
+        // TODO: check that it's an MGLStyleFunction before casting.
+        NSObject *rawValue = toRawStyleSpecValue((MGLStyleFunction<ObjCType> *) value);
+        mbgl::style::conversion::Result<mbgl::style::DataDrivenPropertyValue<MBGLType>> result = mbgl::style::conversion::convert<mbgl::style::DataDrivenPropertyValue<MBGLType>>(rawValue);
+        if (!result) {
+            // raise an exception?
```

Huh -- does this mean that there's no reliable way to fail hard-but-descriptively?  Or are you saying that whenever we _need_ to raise an exception, we should also `NSLog()` so that it isn't maddening to debug?